frontend/src/main/java/com/google/tripmeout/frontend/storage/InMemoryTripModelStorage.java

```diff
@@ -0,0 +1,85 @@
+package com.google.tripmeout.frontend.storage;
+
+import com.google.tripmeout.frontend.TripModel;
+import com.google.tripmeout.frontend.error.TripAlreadyExistsException;
+import com.google.tripmeout.frontend.error.TripNotFoundException;
+import java.util.List;
+import java.util.ArrayList;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+public class InMemoryTripModelStorage {
+  Map<String, TripModel> storage = new ConcurrentHashMap<>();
+
+  public void addTrip(TripModel trip) throws TripAlreadyExistsException {
+    synchronized(storage) {
+      if (storage.get(trip.id()) == null) {
+        storage.put(trip.id(), trip);
+      } else {
+        throw new TripAlreadyExistsException("Trip with id: " + trip.id() + 
+            " already exisits");
+      }
+    }
+  }
+
+  public void removeTrip(String tripId) throws TripNotFoundException {
+    TripModel trip = storage.remove(tripId);
+    if (trip == null) {
+      throw new TripNotFoundException("Trip with id: " + tripId + 
+          " not found in storage");
+    }
+  }
+
+  public void updateTripLocation(String tripId, double latitude, double longitude) 
+      throws TripNotFoundException {
+    synchronized(storage) {
+      TripModel trip = storage.get(tripId);
+      if (trip == null) {
+        throw new TripNotFoundException("Trip with id: " + tripId + 
+            " not found in storage");
+      }
+      TripModel updatedTrip = trip.toBuilder()
+          .setLocationLat(latitude)
+          .setLocationLong(longitude)
+          .build();
+
+      storage.put(tripId, updatedTrip);
+    }
+  }
+
+  public void updateTripName(String tripId, String name) throws TripNotFoundException {
+    synchronized(storage) {
+      TripModel trip = storage.get(tripId);
+      if (trip == null) {
+        throw new TripNotFoundException("Trip with id: " + tripId + 
+            " not found in storage");
+      }
+      TripModel updatedTrip = trip.toBuilder()
+          .setName(name)
+          .build();
+
+      storage.put(tripId, updatedTrip);
+    }
+  }
+
+  public TripModel getTrip(String tripId) throws TripNotFoundException {
+    synchronized(storage) {
+      TripModel trip = storage.get(tripId);
+      if (trip != null) {
+        return trip;
+      }
+      throw new TripNotFoundException("Trip with id: " + tripId + 
+          " not found in storage");
+    } 
+  }
+
+  public List<TripModel> getAllUserTrips(String userId) {
+    List<TripModel> trips = new ArrayList<>();
+    storage.forEach((tripId, trip) -> {
+      if (userId.equals(trip.userId())) {
```

Comment:
The fact that you have to iterate through the map and check for individual matches here shows that a singly-keyed map is probably not the ideal choice here (likely why Chris recommended using a table earlier). There are options such as storing an additional map indexed by user id in order to make these lookups quick. However, that may be overkill since we only intend to use this for testing (as of now; that could change later). (Note that using a second map adds some subtlety with respect to how you do synchronization, but it's still manageable.)

More importantly, consider what behavior you want here when the underlying storage is updated while you iterate over trips here. You probably don't want a `ConcurrentModificationException` or undefined behavior. In that case, you'll need to synchronize on the map in the same way all other accesses are done. There will be performance implications since the entire map will be locked while you do this iteration, but it's the safer approach. For this test-oriented storage system, we expect the data to be small and accesses to be fast, so this probably doesn't matter.

If we did want to support fast access and safe concurrency, the best bet as of now would likely be a second map index, accessed with synchronization. (Better yet, using [persistent data structures](https://en.wikipedia.org/wiki/Persistent_data_structure) which would minimize the need for external locking.) There's also the option of [optimistic locking](https://en.wikipedia.org/wiki/Optimistic_concurrency_control), in which case you assume that concurrent modifications are infrequent and you can somehow roll back any changes and bubble an error up to the client (perhaps to be retried). In this latter case, you _might_ actually want a `ConcurrentHashMap`, which would throw an exception in the bad case without corrupting data.

Anyway, there are a lot of subtleties and design decisions here that it's good to be aware of. For now (depending on your use cases), it's probably best to wrap this in a `synchronized` block and call it a day.

Reply:
One more point to add: if you decide to use optimistic concurrency above, you'll need to use [`ConcurrentHashMap.compute`](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ConcurrentHashMap.html#compute-K-java.util.function.BiFunction-) to perform atomic read-write updates above. Lone `get` calls can be left bare.

Not suggesting you do that here, just pointing out the option.